### ISSUE TITLE
Centralize Exceptions

### ISSUE THREAD

**Issue by itdependsnetworks** [aerleon/aerleon#239]
I think that the custom Exception classes should be centralized, note the following:

```
>>> from aerleon.lib.juniper import UnsupportedFilterError
>>> from aerleon.lib.iptables import UnsupportedFilterError as _UnsupportedFilterError
>>> a = UnsupportedFilterError
>>> a = UnsupportedFilterError()
>>> isinstance(a, UnsupportedFilterError)
True
>>> isinstance(a, _UnsupportedFilterError)
False
>>> from aerleon.lib.iptables import Error
>>> isinstance(a, Error)
False
>>> from aerleon.lib.juniper import Error
>>> isinstance(a, Error)
True
>>> 
```

In here you will note that the re-use of UnsupportedFilterError and Error do not likely follow the Principal of Least Astonishment (POLA). It is very difficult to understand what "Error" you are catching, when that is redefined with the same name in many places. Additionally, it fundamentally makes it difficult for other developers to catch errors from Aerleon. 

I am advocating for something like creating an `errors.py` file and having Exceptions like:

```python
class AerleonException(Exception):

class TermError(AerleonException):
class DuplicateTermError(TermError):

# Cisco Errors
class CiscoTermError(TermError):
class CiscoDuplicateTermError(DuplicateTermError):
```

This way, developers can catch all known Aerleon Errors like:
```python
try:
    do_magic()
except AerleonException:
    pass
```

Or some other generic error like 

```python
try:
    do_magic()
except CiscoTermError:
    pass
```

**Comment by ankenyr** [comment]
Is the intent here to centralize even the generator specific errors? What happens with plugins that people may make and possible release? I like centralizing errors that are common across plugins and having a base exception that all exceptions need to be inherited from.

**Comment by itdependsnetworks** [comment]
> Is the intent here to centralize even the generator specific errors?

I think so. 

> What happens with plugins that people may make and possible release?

Not sure what the issue you are implying is here

**Comment by ankenyr** [comment]
Sorry I should have been more explicit.

I am assuming in `errors.py` you would have the errors for Juniper, Cisco, IPTables, etc and there would be no created errors inside `juniper.py` `cisco.py`, or `iptables.py` etc? If that is the case it ends up being a bit weird. 

Where do we put the errors for [hp-comware](https://github.com/aerleon/aerleon-hp-comware) which is a plugin outside of the main project?

If a user is building automation now they need to understand if a certain generator is a core generator and thus look in `errors.py` or a external one and look in the code/documentation for that.

I like the idea of centralizing the errors that are shared across families such as `AerleonException` `TermError` and `DuplicateTermError` but I am not seeing the benefit of centralizing `CiscoTermError` and similar.

**Comment by itdependsnetworks** [comment]
I guess, it’s the pattern I have seen most pervasive. I can’t think of a single Python lib that doesn’t have either an errors.py or exceptions.py with all exception handling centralized. 

I would guess that someone extending the library would expect the same as well. I know it was confusing to me when I started.

**Comment by jtwb** [comment]
This is definitely a step in the right direction. I agree that it would be helpful to have some common base errors for common cases that can be shared or subclassed by various generators.

> This way, developers can catch all known Aerleon Errors like:
> ```
> try:
>     do_magic()
> except AerleonException:
>     pass
> ```

As people start to write generators plug-ins that we don't necessarily control, it won't be possible to ensure every plug-in subclasses AerleonException. So while we do want to push generator authors towards "the right way" (and I think we should have documentation explaining what is expected when you write a generator), developers should approach error handling with some caution since it only takes one bad plugin to break this assumption.

**Comment by itdependsnetworks** [comment]
Yea, good point, best practices for extending should be documented. Additionally, we can document that we only ensure the provided exceptions come from what core is providing and ideally (but not enforced) what extensions do.

**Comment by ankenyr** [comment]
I have warmed up to the idea. @itdependsnetworks did you want to take a stab at this?

**Comment by itdependsnetworks** [comment]
Yea, that would be great!

**Comment by nemith** [comment]
Taxonomies of types is not a great idea.  I think capirca/aerleon goes a bit overboard with the number of exceptions types and really, when/if there are breaking changes then there should be some proper exception hierarchy and limit most generators to 1 or 2 exception types (or maybe even none).

However just moving them into a separate file doesn't seem beneficial and seems like a anti-pattern.

As for "broken plugins" that shouldn't be much of a concern.  Document how to do it and if a plugin doesn't follow it then it's a quality issue with the plugin.

**Comment by jtwb** [comment]
Agreed with the above. This is an opportunity to reduce the number and nestedness of the exceptions that exist, specifically exceptions that can be raised by generators.

I think we can find a solution with the most value for developers by looking at what kind of exception handlers we want to see upstream of our generators. It is useful to have a central exception class that represents all application-level errors that generators might raise. Callers may reasonably want to catch that exception. In the original proposal this is called "AerleonException", but if we are thinking purely about things that a generator can raise, we might be better off with simply "GeneratorError".

DuplicateTermError is a good test case. Ideally all generators should handle this case in a way that allows us to offer a single user-facing message. At the moment you can see that generators either define their own version of this error (which do not inherit from DuplicateTermError) and each provide their own user message (which varies).

```
aerleon/lib/arista_tp.py:                        raise aclgenerator.DuplicateTermError(
aerleon/lib/cisco.py:                        raise CiscoDuplicateTermError('You have a duplicate term: %s' % term.name)
aerleon/lib/iptables.py:                    raise aclgenerator.DuplicateTermError(
aerleon/lib/juniper.py:                        raise JuniperDuplicateTermError(
aerleon/lib/junipersrx.py:                    raise SRXDuplicateTermError('You have a duplicate term: %s' % term.name)
aerleon/lib/nsxv.py:                    raise NsxvDuplicateTermError('There are multiple terms named: %s' % term.name)
aerleon/lib/packetfilter.py:                    raise DuplicateTermError('You have a duplicate term: %s' % term.name)
aerleon/lib/paloaltofw.py:                    raise PaloAltoFWDuplicateTermError("You have a duplicate term: %s" % term.name)
aerleon/lib/pcap.py:                    raise aclgenerator.DuplicateTermError(
aerleon/lib/windows.py:                    raise aclgenerator.DuplicateTermError(
```

The original proposal does include keeping errors like CiscoXYZ and JuniperXYZ - I agree these are not very useful. It's hard to imagine a caller invoking a generator and then specifically trying to catch a Cisco vs Juniper exception, let alone trying to catch and react to a very specific type of error that the Juniper generator might raise.

> and limit most generators to 1 or 2 exception types (or maybe even none).

My vote would be none. Applications invoking a generator should expect a GeneratorError or no error. Generators should primarily communicate what went wrong through the error message. In the unusual case that a generator needs to provide information that the caller can react to about the error, add data to the error. Some common cases like DuplicateTermError can inherit from GeneratorError in order to have a single error message. But outside of that I would like to see all generator-specific classes removed.

As far as placement goes, there's some debate in this thread about keeping exceptions close to the code that generates them or having a central file for exceptions. My suggestion would be to focus on exceptions that can be raised by generators. Since we need generator authors to know about them and use them, `aclgenerator.py` would be the right place to put them. `aclgenerator.py` acts as kind of a toolkit for generator authors.

**Comment by nemith** [comment]
This entire talk is excellent and i have rewatched it many times but there is a part on namespacing and names that specifically deal with exceptions that i think is very relevant here.

https://www.youtube.com/watch?v=o9pEzgHorH0&t=1231s
